File: sdks/node/src/schedule.ts
/**
 * Schedule handle for managing cron schedules
 */

import type { Reseolio } from './client';
import type { Schedule, ScheduleStatus, JobOptions } from './types';
import { protoToScheduleStatus } from './types';

/**
 * A handle to a cron schedule
 * 
 * Provides methods to manage the schedule lifecycle:
 * - pause/resume to control triggering
 * - update to modify the cron expression, timezone, or options
 * - delete to soft-delete the schedule
 */
export class ScheduleHandle {
    constructor(
        public readonly id: string,
        public readonly name: string,
        private readonly client: Reseolio
    ) { }

    /**
     * Get the current schedule details
     */
    async details(): Promise<Schedule> {
        return this.client.getSchedule(this.id);
    }

    /**
     * Get the current schedule status
     */
    async status(): Promise<ScheduleStatus> {
        const schedule = await this.client.getSchedule(this.id);
        return schedule.status;
    }

    /**
     * Pause the schedule (stops triggering)
     */
    async pause(): Promise<Schedule> {
        return this.client.pauseSchedule(this.id);
    }

    /**
     * Resume a paused schedule
     */
    async resume(): Promise<Schedule> {
        return this.client.resumeSchedule(this.id);
    }

    /**
     * Update the schedule's cron expression, timezone, or options
     */
    async update(options: {
        cron?: string;
        timezone?: string;
        handlerOptions?: JobOptions;
    }): Promise<Schedule> {
        return this.client.updateSchedule(this.id, options);
    }

    /**
     * Delete (soft-delete) the schedule
     */
    async delete(): Promise<boolean> {
        return this.client.deleteSchedule(this.id);
    }

    /**
     * Get the next run time (in milliseconds since epoch)
     */
    async nextRunAt(): Promise<Date> {
        const schedule = await this.client.getSchedule(this.id);
        return new Date(schedule.nextRunAt);
    }

    /**
     * Get the last run time (in milliseconds since epoch)
     */
    async lastRunAt(): Promise<Date | null> {
        const schedule = await this.client.getSchedule(this.id);
        return schedule.lastRunAt ? new Date(schedule.lastRunAt) : null;
    }
}

/**
 * Helper function to convert proto schedule to client format
 */
export function protoToSchedule(proto: any): Schedule {
    return {
        id: proto.id,
        name: proto.name,
        cronExpression: proto.cronExpression,
        timezone: proto.timezone || 'UTC',
        status: protoToScheduleStatus(proto.status),
        nextRunAt: Number(proto.nextRunAt),
        lastRunAt: proto.lastRunAt ? Number(proto.lastRunAt) : 0,
        createdAt: Number(proto.createdAt),
        updatedAt: Number(proto.updatedAt),
        handlerOptions: proto.handlerOptions ? {
            maxAttempts: proto.handlerOptions.maxAttempts,
            backoff: proto.handlerOptions.backoff,
            initialDelayMs: proto.handlerOptions.initialDelayMs,
            maxDelayMs: proto.handlerOptions.maxDelayMs,
            timeoutMs: proto.handlerOptions.timeoutMs,
            jitter: proto.handlerOptions.jitter,
        } : undefined,
    };
}

File: sdks/rust/src/schedule.rs
use crate::{error::Result, types::*};
use chrono::{DateTime, Utc};
use std::sync::Arc;

/// Handle to a cron schedule
pub struct ScheduleHandle {
    id: String,
    name: String,
    client: Arc<crate::client::ReseolioInner>,
}

impl ScheduleHandle {
    pub(crate) fn new(id: String, name: String, client: Arc<crate::client::ReseolioInner>) -> Self {
        Self { id, name, client }
    }

    /// Get schedule ID
    pub fn id(&self) -> &str {
        &self.id
    }

    /// Get schedule name
    pub fn name(&self) -> &str {
        &self.name
    }

    /// Get current schedule details
    pub async fn details(&self) -> Result<Schedule> {
        self.client.get_schedule(&self.id).await
    }

    /// Get current status
    pub async fn status(&self) -> Result<ScheduleStatus> {
        let schedule = self.details().await?;
        Ok(schedule.status)
    }

    /// Pause the schedule
    pub async fn pause(&self) -> Result<Schedule> {
        self.client.pause_schedule(&self.id).await
    }

    /// Resume the schedule
    pub async fn resume(&self) -> Result<Schedule> {
        self.client.resume_schedule(&self.id).await
    }

    /// Update schedule options
    pub async fn update(&self, options: UpdateScheduleOptions) -> Result<Schedule> {
        self.client.update_schedule(&self.id, options).await
    }

    /// Delete (soft-delete) the schedule
    pub async fn delete(&self) -> Result<bool> {
        self.client.delete_schedule(&self.id).await
    }

    /// Get next run time
    pub async fn next_run_at(&self) -> Result<DateTime<Utc>> {
        let schedule = self.details().await?;
        Ok(DateTime::from_timestamp_millis(schedule.next_run_at).unwrap_or_default())
    }

    /// Get last run time
    pub async fn last_run_at(&self) -> Result<Option<DateTime<Utc>>> {
        let schedule = self.details().await?;
        Ok(schedule
            .last_run_at
            .map(|ts| DateTime::from_timestamp_millis(ts).unwrap_or_default()))
    }
}
